src/main/java/it/pagopa/pn/delivery/svc/AttachmentService.java

```diff
@@ -152,52 +145,96 @@ private NotificationAttachment updateSavedAttachment( NotificationAttachment att
     private String saveOneAttachmentToFileStorage( String key, NotificationAttachment attachment ) {
 
         Map<String, String> metadata = new HashMap<>();
-        metadata.put(CONTENT_TYPE_METADATA_NAME, attachment.getContentType() );
         metadata.put(SHA256_METADATA_NAME, attachment.getDigests().getSha256() );
 
         byte[] body = Base64.getDecoder().decode( attachment.getBody() );
 
-        return fileStorage.putFileVersion( key, new ByteArrayInputStream( body ), body.length, metadata );
+        return fileStorage.putFileVersion( key, new ByteArrayInputStream( body ), body.length, attachment.getContentType(), metadata );
+    }
+
+    private NotificationAttachment updateSavedAttachment( NotificationAttachment attachment, String versionId, String fullKey ) {
+        return attachment.toBuilder()
+                .ref( NotificationAttachment.Ref.builder()
+                        .key( fullKey )
+                        .versionToken( versionId )
+                        .build()
+                )
+                .build();
     }
 
+    private void checkAttachmentDigests(NotificationAttachment attachment) {
+        NotificationAttachment.Ref attachmentRef = attachment.getRef();
 
+        try {
+            FileData fd = fileStorage.getFileVersion(
+                    attachmentRef.getKey(), attachmentRef.getVersionToken() );
 
+            String actualSha256 = DigestUtils.sha256Hex( fd.getContent() );
+            validator.checkPreloadedDigests(
+                    attachmentRef.getKey(),
+                    attachment.getDigests(),
+                    NotificationAttachment.Digests.builder()
+                            .sha256( actualSha256 )
+                            .build()
+                );
 
-    public ResponseEntity<Resource> loadDocument(String iun, int documentIndex, String savedVersionId) {
-        String documentKey = String.format("documents/%d", documentIndex );
-        return loadAttachment( iun, documentKey, savedVersionId );
+        } catch (IOException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)